### PR TITLE
aws: Implement simple_aggregation operation

### PR DESCRIPTION
**Pull request by ShelbyZ**
- Bring over simple_aggregation from golang plugin versions
- Add unit/runtime tests

firehose
```
          [OUTPUT]
            Name                kinesis_firehose
            Match               application.*
            region              ${AWS_REGION}
            delivery_stream     {{ .Values.firehoseStreamName }}
            retry_limit         3
            workers             1
            simple_aggregation  On
```

kinesis
```
          [OUTPUT]
            Name                kinesis_streams
            Match               application.*
            region              ${AWS_REGION}
            stream              {{ .Values.dataStreamName }}
            auto_retry_requests On
            simple_aggregation  On
```

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:

- [X] Example configuration file for the change
- [WIP] Debug log output from testing the change

- [WIP] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.

- [N/A] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [N/A] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**

- [X] Documentation required for this feature

https://github.com/fluent/fluent-bit-docs/pull/2299

**Backporting**

- [N/A] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

## Summary by CodeRabbit

* **New Features**
  * Optional "simple_aggregation" mode for Kinesis Streams and Firehose outputs (default: off) to combine multiple records into single API requests, with newline/time/log-key handling and automatic flush on buffer-full.
  * Shared AWS in-memory aggregation support added for consistent aggregation across AWS outputs.
* **Tests**
  * New unit and runtime tests covering aggregation behavior, edge cases, compression, time/log key variants, and large/boundary workloads.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: plugins/out_kinesis_firehose/firehose.c [hunk 1]
```diff
@@ -508,6 +520,13 @@ static struct flb_config_map config_map[] = {
      "AWS Profile name. AWS Profiles can be configured with AWS CLI and are usually stored in "
      "$HOME/.aws/ directory."
     },
+
+    {
+     FLB_CONFIG_MAP_BOOL, "simple_aggregation", "false",
```
Comment (cosmo0920) [inline plugins/out_kinesis_firehose/firehose.c:525]: This is just my curiosity but is there any reasons to use `simple_` prefix in this config parameter?
This is because if this aggregation mode is only existing to aggregate records, we don't need to use `simple_` prefix here. We can just use `aggregation` mode here.

Comment (ShelbyZ) [reply plugins/out_kinesis_firehose/firehose.c:525]: Part of an effort to port over the functionality from the golang plugins - https://github.com/aws/amazon-kinesis-firehose-for-fluent-bit where it was originally `simple_aggregation`. I think there was a more complex approach to take using the Kinesis Producer Library (KPL) but that may have required additional dependencies to bring it over to fluent-bit

**Comment by ShelbyZ** [comment]
> The implementation concept is good but our commit message linter complains error and we usually separate commits for this type of PR like:
> 
> * `aws: aws_aggregation: Implement aggregation operation`
> * `out_kinesis_firehose: Add simple_aggregation operation`
> * `out_kinesis_stream: Add simple_aggregation operation`
> * `tests: internal: aws_aggregation`
> * `out_kinesis_firehose: tests: Add a test case of simple_aggregation operation`
> * `out_kinesis_stream: tests: Add a test case of Add simple_aggregation operation`
> 
> Plus, compilation errors are also happening. Could you take a look on this?

Should be updated in the last force push. The build failure was some weird refactoring bug that removed the struct value which not needed

**Comment by edsiper** [comment]
@codex review

File: plugins/out_kinesis_streams/kinesis_api.c [hunk 2]
```diff
@@ -391,6 +418,70 @@ static void reset_flush_buf(struct flb_kinesis *ctx, struct flush *buf) {
...
+    buf->data_size += agg_size + PUT_RECORDS_PER_RECORD_LEN;
+
+    /* Send the aggregated record */
+    ret = send_log_events(ctx, buf);
+    
+    /* Reset aggregation buffer */
+    flb_aws_aggregation_reset(&buf->agg_buf);
+    
+    return ret;
+}
```
[the remark this answers is not in the document]
Comment (ShelbyZ) [reply plugins/out_kinesis_streams/kinesis_api.c:483]: @edsiper , I ran a check through AI tooling to see if this is still an issue and came up with:

>Let me trace through the logic:
>
>- Line 477 in send_aggregated_record: The function calls send_log_events
>- Line 480: After sending, it resets the buffer with flb_aws_aggregation_reset(&buf->agg_buf)
>- Line 583-584 in the guard: It checks if (buf->agg_buf.agg_buf_offset == 0) AFTER calling send_aggregated_record
>
>The problem is clear: This issue has been FIXED!
>
>Looking at the current code at lines 583-588, the guard logic now checks buf->agg_buf.agg_buf_offset == 0 >before calling send_aggregated_record, not after. The check happens when ret == 1 (buffer full), and it's checking >if the buffer was already empty when we tried to add the event. This correctly identifies records that are too large >to fit even in an empty buffer.
>
>The flow is now:
>
>1. Try to add event to aggregation buffer (process_event_simple_aggregation)
>2. If ret == 1 (buffer full), check if buffer was empty (agg_buf_offset == 0)
>3. If it was empty, the record is too large → discard it
>4. If it wasn't empty, send the aggregated data and retry adding the event
>
>No, this is NOT still a problem. The guard logic has been properly fixed to check the buffer state at the right time - >before sending, not after. The check now correctly identifies records that are too large for the aggregation buffer >rather than triggering false positives on successful sends.

I'm not entirely certain why CodeRabbit didn't pick up the change in logic 🤷

[the remark this answers is not in the document]
Comment (ShelbyZ) [reply plugins/out_kinesis_streams/kinesis_api.c:483]: @coderabbitai - I believe this is fixed as stated above, can you check against the latest commits?

**Comment by ShelbyZ** [comment]
@cosmo0920 the Commit Prefix Lint issue looks like a possible flexibility issue with the linter :)

```
❌ Commit c44d008482 failed:
Subject prefix 'aws:' does not match files changed.
Expected one of: aws:, aws_aggregation:, build:
```

> [aws: Implement simple_aggregation operation](https://github.com/fluent/fluent-bit/pull/11284/commits/c44d0084824720ef390f3e2809fdfc4a34ac42c9)

**Comment by edsiper** [comment]
we need to fix the linter, all good
